pinot-controller/src/main/java/org/apache/pinot/controller/util/BrokerServiceHelper.java

```diff
@@ -0,0 +1,116 @@
+    _controllerConf = controllerConf;
+    _executorService = executor;
+    _connectionManager = connectionManager;
+  }
+
+  @VisibleForTesting
+  public void setCompletionServiceHelper(CompletionServiceHelper completionServiceHelper) {
+    _completionServiceHelper = completionServiceHelper;
+  }
+
+  public BiMap<String, String> getBrokerEndpointsForInstance(List<InstanceConfig> instanceConfigs) {
+    String protocol = _controllerConf.getControllerBrokerProtocol();
+    BiMap<String, String> endpointsToInstances = HashBiMap.create(instanceConfigs.size());
+    for (InstanceConfig instanceConfig : instanceConfigs) {
+      String hostName = instanceConfig.getHostName();
+      if (hostName.startsWith(CommonConstants.Helix.PREFIX_OF_BROKER_INSTANCE)) {
+        hostName = hostName.substring(CommonConstants.Helix.BROKER_INSTANCE_PREFIX_LENGTH);
+      }
+      int port =
+          _controllerConf.getControllerBrokerPortOverride() > 0 ? _controllerConf.getControllerBrokerPortOverride()
+              : Integer.parseInt(instanceConfig.getPort());
+      String brokerEndpoint = String.format("%s://%s:%d", protocol, hostName, port);
+      endpointsToInstances.put(brokerEndpoint, instanceConfig.getInstanceName());
+    }
+    return endpointsToInstances;
+  }
+
+  public TimeBoundaryInfo getTimeBoundaryInfo(TableConfig offlineTableConfig) {
+    String offlineTableName = offlineTableConfig.getTableName();
+
+    List<InstanceConfig> instanceConfigs = _pinotHelixResourceManager.getBrokerInstancesConfigsFor(offlineTableName);
+    BiMap<String, String> endpointsToInstances = getBrokerEndpointsForInstance(instanceConfigs);
+
+    CompletionServiceHelper completionServiceHelper;
+    if (_completionServiceHelper == null) {
+      completionServiceHelper = new CompletionServiceHelper(_executorService, _connectionManager, endpointsToInstances);
+    } else {
+      completionServiceHelper = _completionServiceHelper;
+    }
+
+    List<String> timeBoundaryInfoUris = new ArrayList<>(endpointsToInstances.size());
+    for (String endpoint : endpointsToInstances.keySet()) {
+      String timeBoundaryInfoUri = endpoint + String.format(TIME_BOUNDARY_INFO_API_PATH, offlineTableName);
+      timeBoundaryInfoUris.add(timeBoundaryInfoUri);
+    }
+    Map<String, String> reqHeaders = new HashMap<>();
+    CompletionServiceHelper.CompletionServiceResponse serviceResponse =
+        completionServiceHelper.doMultiGetRequest(timeBoundaryInfoUris, offlineTableName, false, reqHeaders,
+            DEFAULT_REQUEST_TIMEOUT_MS, "get time boundary information for table from broker instances");
+    for (Map.Entry<String, String> streamResponse : serviceResponse._httpResponses.entrySet()) {
+      try {
+        return JsonUtils.stringToObject(streamResponse.getValue(), TimeBoundaryInfo.class);
+      } catch (JsonProcessingException e) {
+        LOGGER.error("Failed to parse timeBoundaryInfo response: {}", streamResponse.getKey(), e);
+      }
+    }
+    return null;
```

Comment:
Suggest throwing the exception out instead of using `null` to pass exception. The caller can catch the exception and log the stack trace with more context